pkg/volume/csi/csi_plugin.go

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2014 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"errors"
+	"fmt"
+	"path"
+	"time"
+
+	"github.com/golang/glog"
+	api "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume"
+	csipb "k8s.io/kubernetes/pkg/volume/csi/proto/csi"
+)
+
+const (
+	csiName       = "csi"
+	csiPluginName = "kubernetes.io/csi"
+	//csiAddrTemplate = "/var/lib/kubelet/plugins/csi/sockets/%v/kubeletproxy.sock"
+	csiAddrTemplate = "/tmp/kubeletproxy.sock"
+	csiTimeout      = 15 * time.Second
+)
+
+var (
+	// csiVersion supported csi version
+	csiVersion = &csipb.Version{Major: 0, Minor: 1, Patch: 0}
+)
+
+type csiPlugin struct {
+	host volume.VolumeHost
+}
+
+// ProbeVolumePlugins returns implemented plugins
+func ProbeVolumePlugins() []volume.VolumePlugin {
```

Comment:
ProbeVolumePlugins is ok for now (with hardcoded socket), but it should be called dynamically when kubelet discovers a new CSI driver on the machine. Some TODO comment would be nice.

I assume in future there will be multiple instances of `csiPlugin`, one for each driver (following the Flex model).

Reply:
> multiple instances of csiPlugin, one for each driver 

Well... this is going to be very interesting. There will be **multiple instances** of csiPlugin, one for each driver **only on nodes**. Controller-manager (provisioner, a/d controller) will have **one** instance of the plugin for all CSI drivers - there is no way how to enumerate drivers on master.

It *should* work, however we need to test it carefulyl.

Reply:
@jsafrane agreed, will add TODO.  Discovery is still not clear and we definitely need to come back and careful tests.  Solution can be as simple as a known path for socket (a la flexvolume) or resource registry.